cpp/src/arrow/parquet/writer.cc

```diff
@@ -70,36 +72,124 @@ Status FileWriter::Impl::NewRowGroup(int64_t chunk_size) {
   return Status::OK();
 }
 
-template <typename ParquetType>
+#define CONVERT_PHYSICAL_TYPE_COPY_DATA(InType, OutType)                                \
+  template <>                                                                           \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(                        \
+      const InType* data_ptr, int64_t length, const OutType** out_ptr) {                \
+    RETURN_NOT_OK(data_buffer_.Resize(length * sizeof(OutType)));                       \
+    OutType* mutable_out_ptr = reinterpret_cast<OutType*>(data_buffer_.mutable_data()); \
+    std::copy(data_ptr, data_ptr + length, mutable_out_ptr);                            \
+    *out_ptr = mutable_out_ptr;                                                         \
+    return Status::OK();                                                                \
+  }
+
+#define CONVERT_PHYSICAL_TYPE_COPY_PTR(InType, OutType)           \
+  template <>                                                     \
+  Status FileWriter::Impl::ConvertPhysicalType<InType, OutType>(  \
+      const InType* data_ptr, int64_t, const OutType** out_ptr) { \
+    *out_ptr = reinterpret_cast<const OutType*>(data_ptr);        \
+    return Status::OK();                                          \
+  }
+
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int8_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(uint16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_DATA(int16_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(int32_t, int32_t)
+CONVERT_PHYSICAL_TYPE_COPY_PTR(uint64_t, int64_t)
```

Comment:
Probably nothing different will happen if we just copy and cast these integers. The ones in the range  0<=x<=INT_MAX will be same in both, otherwise the ones >INT_MAX will overflow and return as negative. If both sides support Parquet 2, you should be able to pass all possible uint32 values like this to the reader.